File: terraform/import_lambda.py
#!/usr/bin/env python3
import os
import sys
import json
import zipfile
import boto3
import requests

# Configuration
FUNCTIONS_DIR = "../functions"
TERRAFORM_DIR = "../terraform"
GENERATED_TF_FILE = os.path.join(
    TERRAFORM_DIR, "auto_generated_imported_functions.tf")


def get_lambda_function(function_name):
    client = boto3.client("lambda")
    try:
        response = client.get_function(FunctionName=function_name)
        return response
    except Exception as e:
        print(f"Error retrieving function '{function_name}': {e}")
        sys.exit(1)


def download_code(url, download_path):
    print(f"Downloading code from: {url}")
    response = requests.get(url)
    if response.status_code != 200:
        print(f"Failed to download code: HTTP {response.status_code}")
        sys.exit(1)
    with open(download_path, "wb") as f:
        f.write(response.content)
    print(f"Downloaded code to: {download_path}")


def extract_zip(zip_path, extract_to):
    print(f"Extracting {zip_path} to {extract_to} ...")
    with zipfile.ZipFile(zip_path, "r") as zip_ref:
        zip_ref.extractall(extract_to)
    print("Extraction complete.")


def generate_terraform_file(function_name, config):
    # Basic resource block with some configuration fields from AWS.
    # You can customize this as needed.
    tf_resource = f"""
resource "aws_lambda_function" "{function_name}" {{
  function_name = "{function_name}"
  role          = "{config.get('Role', 'REPLACE_WITH_ROLE_ARN')}"
  handler       = "{config.get('Handler', 'index.handler')}"
  runtime       = "{config.get('Runtime', 'nodejs18.x')}"
  filename      = "${{path.module}}/build/{function_name}.zip"

  // Ensure Terraform notices code changes in the .zip:
  source_code_hash = filebase64sha256("${{path.module}}/build/{function_name}.zip")
"""

    # Optionally add VPC configuration if available
    vpc_config = config.get("VpcConfig")
    if vpc_config:
        subnets = json.dumps(vpc_config.get("SubnetIds", []))
        sgs = json.dumps(vpc_config.get("SecurityGroupIds", []))
        tf_resource += f"""
  vpc_config {{
    subnet_ids         = {subnets}
    security_group_ids = {sgs}
  }}
"""

    # Optionally add environment variables if available
    env = config.get("Environment", {}).get("Variables")
    if env:
        env_json = json.dumps(env)
        tf_resource += f"""
  environment {{
    variables = {env_json}
  }}
"""
    tf_resource += "\n}\n"

    # Optionally, add an output block.
    tf_output = f"""
output "{function_name}_name" {{
  value = aws_lambda_function.{function_name}.function_name
}}
"""

    return tf_resource + tf_output


def main():
    if len(sys.argv) < 2:
        print("Usage: import_lambda.py <FunctionName>")
        sys.exit(1)

    function_name = sys.argv[1]
    print(f"Importing Lambda function: {function_name}")

    # Get the Lambda function details from AWS.
    function_data = get_lambda_function(function_name)
    configuration = function_data.get("Configuration", {})
    code = function_data.get("Code", {})
    code_url = code.get("Location")
    if not code_url:
        print("No code URL found in the function data.")
        sys.exit(1)

    # Create a folder in FUNCTIONS_DIR for the function.
    function_folder = os.path.join(FUNCTIONS_DIR, function_name)
    os.makedirs(function_folder, exist_ok=True)

    # Download the code package.
    zip_file_path = os.path.join(function_folder, f"{function_name}.zip")
    download_code(code_url, zip_file_path)

    # Extract the zip file into the function folder.
    extract_zip(zip_file_path, function_folder)

    # (Optional) Move the zip file to your Terraform build folder.
    build_dir = os.path.join(TERRAFORM_DIR, "build")
    os.makedirs(build_dir, exist_ok=True)
    terraform_zip_path = os.path.join(build_dir, f"{function_name}.zip")
    os.rename(zip_file_path, terraform_zip_path)
    print(f"Moved zip file to Terraform build directory: {terraform_zip_path}")

    # Generate a Terraform configuration for this function.
    tf_block = generate_terraform_file(function_name, configuration)

    # Append to (or create) the auto-generated Terraform file.
    with open(GENERATED_TF_FILE, "a") as f:
        f.write(tf_block)
        f.write("\n")

    print(
        f"Terraform configuration for '{function_name}' generated in '{GENERATED_TF_FILE}'.")
    print("To import this function into Terraform state, run:")
    print(
        f"  terraform import aws_lambda_function.{function_name} {function_name}")


if __name__ == "__main__":
    main()

File: scripts/import_lambda.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const https = require('https');
const { URL } = require('url');
const AWS = require('aws-sdk');
const AdmZip = require('adm-zip');

// Directories and file paths (adjust these paths as needed)
const FUNCTIONS_DIR = path.join(__dirname, '../functions');
const TERRAFORM_DIR = path.join(__dirname, '../terraform');
const BUILD_DIR = path.join(TERRAFORM_DIR, 'build');
// JSON file to hold function-specific configurations
const CONFIG_FILE = path.join(TERRAFORM_DIR, 'lambda_config.json');

function loadConfig() {
    if (fs.existsSync(CONFIG_FILE)) {
        const data = fs.readFileSync(CONFIG_FILE, 'utf8');
        return JSON.parse(data);
    }
    return {};
}

function saveConfig(config) {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(config, null, 2));
}

function updateFunctionConfig(functionName, configData) {
    // Map fields from AWS configuration to our JSON configuration
    const newConfig = {};
    newConfig.runtime = configData.Runtime || 'nodejs18.x';
    newConfig.handler = configData.Handler || 'index.handler';
    newConfig.role = configData.Role || 'REPLACE_WITH_ROLE_ARN';

    if (configData.VpcConfig) {
        newConfig.vpc_config = {
            subnet_ids: configData.VpcConfig.SubnetIds || [],
            security_group_ids: configData.VpcConfig.SecurityGroupIds || []
        };
    }

    const env = configData.Environment && configData.Environment.Variables;
    if (env) {
        newConfig.environment = env;
    }

    if (configData.Layers) {
        newConfig.layers = configData.Layers
            .filter(layer => layer.Arn)
            .map(layer => layer.Arn);
    }

    // Load existing config, update the function entry, and save it back.
    const config = loadConfig();
    config[functionName] = newConfig;
    saveConfig(config);
    console.log(`Updated configuration for function '${functionName}' in ${CONFIG_FILE}.`);
}

async function getLambdaFunction(functionName) {
    AWS.config.update({ region: 'eu-central-1' });
    const lambda = new AWS.Lambda();
    try {
        const response = await lambda.getFunction({ FunctionName: functionName }).promise();
        return response;
    } catch (error) {
        console.error(`Error retrieving function '${functionName}': ${error.message}`);
        process.exit(1);
    }
}

function downloadCode(url, downloadPath) {
    return new Promise((resolve, reject) => {
        console.log(`Downloading code from: ${url}`);
        const parsedUrl = new URL(url);
        const options = {
            hostname: parsedUrl.hostname,
            path: parsedUrl.pathname + parsedUrl.search,
            protocol: parsedUrl.protocol,
        };

        https.get(options, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(`Failed to download code: HTTP ${res.statusCode}`));
                return;
            }
            const fileStream = fs.createWriteStream(downloadPath);
            res.pipe(fileStream);
            fileStream.on('finish', () => {
                fileStream.close();
                console.log(`Downloaded code to: ${downloadPath}`);
                resolve();
            });
        }).on('error', (err) => {
            reject(err);
        });
    });
}

function extractZip(zipPath, extractTo) {
    console.log(`Extracting ${zipPath} to ${extractTo} ...`);
    const zip = new AdmZip(zipPath);
    zip.extractAllTo(extractTo, true);
    console.log('Extraction complete.');
}

async function main() {
    if (process.argv.length < 3) {
        console.error('Usage: import_lambda.js <FunctionName>');
        process.exit(1);
    }

    const functionName = process.argv[2];
    console.log(`Importing Lambda function: ${functionName}`);

    // Retrieve the function's configuration and code details from AWS.
    const functionData = await getLambdaFunction(functionName);
    const configuration = functionData.Configuration || {};
    const code = functionData.Code || {};
    const codeUrl = code.Location;
    if (!codeUrl) {
        console.error('No code URL found in the function data.');
        process.exit(1);
    }

    // Create a folder for the function under FUNCTIONS_DIR.
    const functionFolder = path.join(FUNCTIONS_DIR, functionName);
    if (!fs.existsSync(functionFolder)) {
        fs.mkdirSync(functionFolder, { recursive: true });
    }

    // Download the code package as a ZIP file into the function folder.
    const zipFilePath = path.join(functionFolder, `${functionName}.zip`);
    try {
        await downloadCode(codeUrl, zipFilePath);
    } catch (error) {
        console.error(`Error downloading code: ${error.message}`);
        process.exit(1);
    }

    // Extract the zip file into the function folder.
    extractZip(zipFilePath, functionFolder);

    // Ensure the Terraform build directory exists, then move the ZIP file there.
    if (!fs.existsSync(BUILD_DIR)) {
        fs.mkdirSync(BUILD_DIR, { recursive: true });
    }
    const terraformZipPath = path.join(BUILD_DIR, `${functionName}.zip`);
    fs.renameSync(zipFilePath, terraformZipPath);
    console.log(`Moved zip file to Terraform build directory: ${terraformZipPath}`);

    // Update the configuration JSON file with the function's settings.
    updateFunctionConfig(functionName, configuration);
}

main();
